### ISSUE TITLE
Not working on Android 13 since READ_EXTERNAL_STORAGE can't be request

### ISSUE THREAD

**Issue by bahung1221** [miguelpruivo/flutter_file_picker#1099]
**Describe the bug**
In Android 13 the READ_EXTERNAL_STORAGE has been replaced by READ_MEDIA_IMAGES, READ_MEDIA_VIDEO and READ_MEDIA_AUDIO. So the READ_EXTERNAL_STORAGE permission can't be using anymore.
So I tried to request the READ_MEDIA_IMAGES  permission above and then try to use file_picker but it thrown `read_external_storage_denied` exception.

[https://developer.android.com/about/versions/13/behavior-changes-13](https://developer.android.com/about/versions/13/behavior-changes-13)

**Platform**

- [x] Android
- [ ] iOS
- [ ] Web
- [ ] Desktop

**Platform OS version**
Android 13 (API level 33)

**How are you picking?**
```
FilePickerResult? result = await FilePicker.platform.pickFiles(type: FileType.image);
```

**Details to reproduce the issue**

- AndroidManifest.xml
```xml
<manifest
  xmlns:android="http://schemas.android.com/apk/res/android"
  xmlns:tools="http://schemas.android.com/tools"
  package="com.your.package.example"
>
  <uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
  <uses-permission android:name="android.permission.READ_MEDIA_VIDEO" />
  <uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />
  <uses-permission
    android:name="android.permission.READ_EXTERNAL_STORAGE"
    android:maxSdkVersion="32"
  />
  ...
</manifest>
```

- Request permission (custom Android code because `flutter_permission_handler`library still not updated to support above permissions):
```kt
class MainActivity: FlutterActivity() {
  private val CUSTOM_PERMISSION_HANDLER_CHANNEL = "example/custom_permission_handler"
  private val CUSTOM_PERMISSION_REQUEST_CODE = 101
  private val CUSTOM_PERMISSION_HANDLER_TAG = "CustomPermissionHandler"
  private lateinit var requestingPermission: String
  private lateinit var permissionResultCallback: (Int) -> Unit

  ...

  @RequiresApi(33)
  fun requestPermission() {
    when {
      ContextCompat.checkSelfPermission(
        this,
        Manifest.permission.READ_MEDIA_IMAGES
      ) == PackageManager.PERMISSION_GRANTED -> {
        // Already granted
        Log.i(CUSTOM_PERMISSION_HANDLER_TAG, "ccc")
        permissionResultCallback(1)
      }

      else -> {
        // You can directly ask for the permission.
        // The registered ActivityResultCallback gets the result of this request.
        ActivityCompat.requestPermissions(this,
          arrayOf(Manifest.permission.READ_MEDIA_IMAGES),
          CUSTOM_PERMISSION_REQUEST_CODE
        )
      }
    }
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<String>,
    grantResults: IntArray
  ) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)

    when (requestCode) {
      CUSTOM_PERMISSION_REQUEST_CODE -> {
        if (grantResults.isEmpty() || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
          if (shouldShowRequestPermissionRationale(Manifest.permission.READ_MEDIA_IMAGES)) {
            permissionResultCallback(-1)
            return
          }

          Log.i(CUSTOM_PERMISSION_HANDLER_TAG, "Permission has been denied by user")
          permissionResultCallback(0)
        } else {
          Log.i(CUSTOM_PERMISSION_HANDLER_TAG, "Permission has been granted by user")
          permissionResultCallback(1)
        }
      }
    }
  }
}
```

**Flutter Version details**
```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.3.1 21E258 darwin-arm, locale en-VN)
    • Flutter version 3.0.5 at /Users/m/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (6 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/m/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/m/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64  • Android 13 (API 33) (emulator)
    • macOS (desktop)             • macos         • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 104.0.5112.101
    ! Error: Failed to prepare device for development. Please unlock and reconnect the device. (code
      806)

[✓] HTTP Host Availability
    • All required HTTP hosts are available
```

**Comment by SherwinDeveloper** [comment]
Having the same challenge.  Need to target Android 13 for other features.  Is there a workaround for this?

**Comment by bahung1221** [comment]
@SherwinDeveloper ,

I implemented the custom picker for Android 13 only by Kotlin to work around for this issue:

- **MainActivity.kt**:
```kt
class MainActivity: FlutterActivity() {
  private val CUSTOM_FILE_PICKER_CHANNEL = "com.example/custom_file_picker"
  private val CUSTOM_FILE_PICKER_REQUEST_CODE = 102
  private val CUSTOM_FILE_PICKER_TAG = "CustomPermissionHandler"
  private lateinit var pickPdfResultCallback: (String?) -> Unit

  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
    super.configureFlutterEngine(flutterEngine)

    registerCustomFilePicker(flutterEngine)
  }

  private fun registerCustomFilePicker(@NonNull flutterEngine: FlutterEngine) {
    MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CUSTOM_FILE_PICKER_CHANNEL).setMethodCallHandler {
      call, result ->
      if (call.method == "pickPdf") {
        pickPdfResultCallback = fun (filePath: String?) {
          result.success(filePath)
        }
        val pickFileIntent = Intent(Intent.ACTION_GET_CONTENT)
        pickFileIntent.type = "application/pdf"

        activity.startActivityForResult(pickFileIntent, CUSTOM_FILE_PICKER_REQUEST_CODE)
      }
      else {
        result.notImplemented()
      }
    }
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
    super.onActivityResult(requestCode, resultCode, intent)

    if (requestCode == CUSTOM_FILE_PICKER_REQUEST_CODE) {
      handleChooseFileResult(resultCode, intent)
    }
  }

  private fun handleChooseFileResult(resultCode: Int, intent: Intent?) {
    if (resultCode == RESULT_OK && intent != null) {
      if (intent.data == null) {
        pickPdfResultCallback(null)
        return
      }

      val pathAndName: String? = getPathFromUri(activity, intent.data!!)
      pickPdfResultCallback(pathAndName)
      return
    }

    // User cancelled choosing a file.
    pickPdfResultCallback(null)
  }

  private fun getPathFromUri(context: Context, uri: Uri): String? {
    var file: File? = null
    var originalName: String? = null
    var inputStream: InputStream? = null
    var outputStream: OutputStream? = null
    var success = false
    try {
      val extension: String? = getFileExtension(context, uri)
      originalName = getFileName(context, uri)

      inputStream = context.getContentResolver().openInputStream(uri)
      file = File.createTempFile("custom_file_picker", extension, context.getCacheDir())
      file.deleteOnExit()
      outputStream = FileOutputStream(file)
      if (inputStream != null) {
        copy(inputStream, outputStream)
        success = true
      }
    } catch (ignored: IOException) {
    } finally {
      try {
        if (inputStream != null) inputStream.close()
      } catch (ignored: IOException) {
      }
      try {
        if (outputStream != null) outputStream.close()
      } catch (ignored: IOException) {
        // If closing the output stream fails, we cannot be sure that the
        // target file was written in full. Flushing the stream merely moves
        // the bytes into the OS, not necessarily to the file.
        success = false
      }
    }

    return if (success && file != null) "${file.getPath()},$originalName" else null
  }

  /** @return extension of file with dot
   */
  private fun getFileExtension(context: Context, uriFile: Uri): String? {
    var extension: String? = null
    extension = try {
      if (uriFile.scheme == ContentResolver.SCHEME_CONTENT) {
        val mime = MimeTypeMap.getSingleton()
        mime.getExtensionFromMimeType(context.contentResolver.getType(uriFile))
      } else {
        MimeTypeMap.getFileExtensionFromUrl(
          Uri.fromFile(File(uriFile.path)).toString())
      }
    } catch (e: Exception) {
      null
    }
    if (extension == null || extension.isEmpty()) {
      //default extension for matches the previous behavior of the plugin
      return ""
    }
    return ".$extension"
  }

  /** @return name of file
   */
  fun getFileName(context: Context, uri: Uri): String? {
    var result: String? = null
    try {
      if (uri.scheme == "content") {
        val cursor: Cursor? = context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)
        try {
          if (cursor != null && cursor.moveToFirst()) {
            result = cursor.getString(cursor.getColumnIndexOrThrow(OpenableColumns.DISPLAY_NAME))
          }
        } finally {
          cursor?.close()
        }
      }
      if (result == null) {
        result = uri.path
        val cut = result!!.lastIndexOf('/')
        if (cut != -1) {
          result = result.substring(cut + 1)
        }
      }
    } catch (ex: java.lang.Exception) {
      Log.e(CUSTOM_FILE_PICKER_TAG, "Failed to handle file name: $ex")
    }
    return result
  }

  @Throws(IOException::class)
  private fun copy(`in`: InputStream, out: OutputStream) {
    val buffer = ByteArray(4 * 1024)
    var bytesRead: Int
    while (`in`.read(buffer).also { bytesRead = it } != -1) {
      out.write(buffer, 0, bytesRead)
    }
    out.flush()
  }
}
```

- **CustomFilePicker.dart**:
```dart
import 'dart:convert';
import 'dart:io';

import 'package:cross_file/cross_file.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:file_picker/file_picker.dart';
import 'package:flutter/services.dart';

class CustomFile {
  String fileName;
  String base64;

  CustomFile({
    required this.fileName,
    required this.base64,
  });
}

class CustomFilePicker {
  static const platform = MethodChannel('jp.lnkr/custom_file_picker');

  static Future<CustomFile?> pickPdf() async {
    if (Platform.isIOS) {
      return _pickPdfByLib();
    }

    DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
    AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
    if (androidInfo.version.sdkInt! < 33) {
      return _pickPdfByLib();
    }

    // Custom implementation for android API 33 and higher only
    return _customPickPdf();
  }

  ///
  ///
  ///
  static Future<CustomFile?> _pickPdfByLib() async {
      // .. flutter_file_picker
  }

  ///
  ///
  ///
  static Future<CustomFile?> _customPickPdf() async {
    try {
      final String? pathAndName = await platform.invokeMethod('pickPdf');
      if (pathAndName == null ) {
        return null;
      }

      final List<String> result = pathAndName.split(',');
      final String path = result[0];
      final String? name = result.length > 1 ? result[1] : null;
      final XFile file = XFile(path);

      return CustomFile(
        fileName: name ?? file.name,
        base64: base64.encode(await file.readAsBytes()),
      );
    } on PlatformException catch (e) {
      rethrow;
    }
  }
}
```

**Comment by rlueders** [comment]
This bug makes the package complete worthless on Android since the Android 13 release.  Is this plugin even supported anymore?

**Comment by alexmercerind** [comment]
Can you guys try out the fork at #1120?
I think not requesting runtime permission `READ_EXTERNAL_STORAGE` on SDK 33 or higher (Android 13 or above) just works.

**Comment by bahung1221** [comment]
@alexmercerind ,

I can't try it now, but I think we should request `READ_MEDIA_IMAGES` / `READ_MEDIA_VIDEO` / `READ_MEDIA_AUDIO` permissions when the pick type is `FileType.image` / `FileType.video` / `FileType.audio`.

**Comment by alexmercerind** [comment]
@bahung1221 ,

I'm using my fork in my project and just removing permission for Android 13 works.
READ_MEDIA_IMAGES / READ_MEDIA_VIDEO / READ_MEDIA_AUDIO are only for accessing content through `MediaStore`.
When user themselves select a file from file picker, it inherently doesn't require any permission.

Again, I'm just looking for confirmation.

**Comment by bahung1221** [comment]
@alexmercerind ,

Sorry for the misleading, I tried your patch and it worked well on my project for all file types.

```yaml
  # file_picker: ^4.6.1
  file_picker:
    git:
      url: https://github.com/alexmercerind/flutter_file_picker.git
      ref: master
```

Thanks a lot! 🍻

**Comment by aliyazdi75** [comment]
@miguelpruivo I still can reproduce this issue on Android 13:
```======== Exception caught by image resource service ================================================
The following FileSystemException was thrown resolving an image codec:
Cannot open file, path = '/storage/emulated/0/Pictures/IMG_20220923_110941.jpg' (OS Error: Permission denied, errno = 13)
```

**Comment by aliyazdi75** [comment]
> @aliyazdi75, what's your `compileSdkVersion` & `targetSdkVersion`? Can you try with `5.1.0` pinned in your `pubspec.yaml`?

They both are 33. You mean to downgrade it to `5.1.0`?

**Comment by alexmercerind** [comment]
Yes. Since #1123 brought some big changes to Android native implementation.

**Comment by aliyazdi75** [comment]
@alexmercerind Yes, worked thanks. I think they should revert those broken changes.

**Comment by abhi16180** [comment]
> @alexmercerind Yes, worked thanks. I think they should revert those broken changes.
> 
> CC: @abhi16180
@aliyazdi75 
I am not sure the if the issue is due to the changes  that made in v5.2.0 because I haven't touched FilePickerDelegate.java file.
The last update is about preventing unnecessary caching. Only FileUtils.java was changed.

**Comment by alexmercerind** [comment]
Likely @aliyazdi75 is using the `dart:io` `File` class on the path returned by the `package:file_picker` to read it (from the logs).

As far as I see @abhi16180's implementation prevents caching by not caching the files which are media (can be accessed by just `READ_EXTERNAL_STORAGE` even after scoped-storage enforcement), but for non-media files it actually caches stuff.

Here's why that won't work after Android 13:

In newer Android versions (Android 10 or higher, scoped storage enforced), when picking a file using the native picker `ACTION_GET_CONTENT`, it inherently gives permission to the file/URI... BUT, that doesn't mean you can access it using it's `File` path only through the `InputStream` or `FileDescriptor`, both of which have no equivalent in Dart.
One more thing, the URI returned by file-picker doesn't require any permission at all actually (!!!), only if you access it through `InputStream`.

The Dart's `dart:io` `File` class obviously doesn't make use of `InputStream` & is likely backed by `java.io.File`.
Now one can ask, then why does the `File` class from `dart:io` works on cached files? That's because you have complete ownership & access to files in your directory (`/storage/emulated/0/Android/data/com.abc.xyz/files`).

If you want to avoid caching media files when targeting Android 13, you'll have to request those [3 new permissions](https://developer.android.com/about/versions/13/behavior-changes-13#granular-media-permissions) and it'll require 3 permission dialogs just before opening a file-picker (for the first time).

**Comment by abhi16180** [comment]
> Likely @aliyazdi75 is using the `dart:io` `File` class on the path returned by the `package:file_picker` to read it (my guess, from the logs).
> 
> As far as I see @abhi16180's implementation prevents caching by not caching the files which are media (can be accessed by just `READ_EXTERNAL_STORAGE` even after scoped-storage enforcement), but for non-media files it actually caches stuff.
> 
> Here's why that won't work after Android 13:
> 
> In newer Android versions (Android 10 or higher, scoped storage enforced), when picking a file using the native picker `ACTION_GET_CONTENT`, it inherently gives permission to the file/URI... BUT, that doesn't mean you can access it using it's `File` path only through the `InputStream` or `FileDescriptor`, both of which have no equivalent in Dart. One more thing, the URI returned by file-picker doesn't require any permission at all actually (!!!), only if you access it through `InputStream`.
> 
> The Dart's `dart:io` `File` class obviously doesn't make use of `InputStream` & is likely backed by `java.io.File`. Now one can ask, then why does the `File` class from `dart:io` works on cached files? That's because you have complete ownership & access to files in your directory (`/storage/emulated/0/Android/data/com.abc.xyz/files`).
> 
> If you want to avoid caching media files when targeting Android 13, you'll have to request those [3 new permissions](https://developer.android.com/about/versions/13/behavior-changes-13#granular-media-permissions) because `READ_EXTERNAL_STORAGE` no longer works, and it'll require 3 permission dialogs just before opening a file-picker (for the first time).

#1120 

@alexmercerind 
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            if (!this.permissionManager.isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE)) {
                this.permissionManager.askForPermission(Manifest.permission.READ_EXTERNAL_STORAGE, REQUEST_CODE);
                return;
            }
        }
        
        
      This doesn't request for READ_EXTERNAL_STORAGE for android 13 or above, in the else case if we request 
      - READ_MEDIA_IMAGES
      - READ_MEDIA_VIDEOS
      - READ_MEDIA_AUDIO
      
      It should work right?

**Comment by abhi16180** [comment]
@aliyazdi75 @alexmercerind 
I have added 
- READ_MEDIA_IMAGES
- READ_MEDIA_VIDEO
- READ_MEDIA_AUDIO 
permissions.
Can you please try and tell me if this is working you ? 
[android-13-fix](https://github.com/abhi16180/flutter_file_picker.git)

**Comment by aliyazdi75** [comment]
I did yesterday on my side but no it didn't work.

**Comment by abhi16180** [comment]
> I did yesterday on my side but no it didn't work.

Oh, actually my app uses the same thing but it worked on Android 13(I got feedback from user). I don't have Android 13, so it's kinda difficult for me to reproduce the error that you are getting.

**Comment by abhi16180** [comment]
@aliyazdi75  I have tried it on emulator (sdk 33), I don't have physical device to test it . But it is working properly here , Can you check it out ?
![uncached](https://user-images.githubusercontent.com/64468853/192102806-b55b5c73-070b-415b-a203-c0715b9aa86b.png)
![cached](https://user-images.githubusercontent.com/64468853/192102809-05cd5a21-cc73-4fbc-a454-fe9e1920aff3.png)

**Comment by abhi16180** [comment]
> @aliyazdi75 I have tried it on emulator (sdk 33), I don't have physical device to test it . But it is working properly here , Can you check it out ? ![uncached](https://user-images.githubusercontent.com/64468853/192102806-b55b5c73-070b-415b-a203-c0715b9aa86b.png) ![cached](https://user-images.githubusercontent.com/64468853/192102809-05cd5a21-cc73-4fbc-a454-fe9e1920aff3.png)

@aliyazdi75 
@alexmercerind 
And without any modification it is working fine on emulator (android 13). I am unable to reproduce the problem. (Using v5.2.0)

**Comment by alexmercerind** [comment]
@abhi16180,

> And even without any modification it is working fine on emulator (android 13). I am unable to reproduce the problem. (Using v5.2.0)

I think the problem is that The Dart's dart:io `File` class is not working on the non-cached returned paths (in @ aliyazdi75's logs a `FileSystemException` is thrown, which is a class from Dart itself).
Unfortunately, I'm still using my v5.1.0 fork in production, so can't confirm.

**Comment by abhi16180** [comment]
> @abhi16180,
> 
> > And even without any modification it is working fine on emulator (android 13). I am unable to reproduce the problem. (Using v5.2.0)
> 
> I think the problem is that The Dart's dart:io `File` class is not working on the non-cached returned paths.
> 
> In @ aliyazdi75's logs, a `FileSystemException` is thrown, which is a class from Dart itself. I'm talking about without permission scenario (current v5.2.0 tree). `OS Error: Permission denied, errno = 13`.
> 
> Unfortunately, I'm still using my v5.1.0 fork in production, so can't confirm.
> 
> I haven't read the complete API but maybe `package:file_picker` already returns `bytes` of the selected files. So, maybe we can still prevent caching, by restricting access to file paths & marking it an breaking-API change. Some other people at #1133 seem to be facing some other issue, I'm not Android developer by any means, I don't know if `MediaStore` queries with DATA column are a safe way to rely on. No idea.

My app uses modified version of file-picker. Later I sent PR and it got merged (v5.2.0). I don't think "dart:io" doesn't work on uncached file paths. Because I was able to use run my app without any issue on android 11. And when it comes to android 13 , I cannot test it on physical device. But on emulator it didn't show any issue.

**Comment by alexmercerind** [comment]
> I don't think "dart:io" doesn't work on un-cached file paths.

It won't work because you don't have the "file access" to the file through `READ_EXTERNAL_STORAGE` or any of `READ_MEDIA_IMAGES`, `READ_MEDIA_VIDEO` or `READ_MEDIA_AUDIO`.

Having these permissions means, you can read it through:
- `MediaStore` APIs.
- Creating a `FileDescriptor` and using `InputStream`.

Access through file path will STILL give permission error (on any SDK level higher than 29, Android 10).

Similar instances may be found at: https://stackoverflow.com/questions/47125837/java-io-filenotfoundexception-from-intent-data-tostring.

**Comment by abhi16180** [comment]
> > I don't think "dart:io" doesn't work on un-cached file paths.
> 
> It won't work because you don't have the "file access" to the file through `READ_EXTERNAL_STORAGE` or any of `READ_MEDIA_IMAGES`, `READ_MEDIA_VIDEO` or `READ_MEDIA_AUDIO`.
> 
> Having these permissions means, you can read it through:
> 
> * `MediaStore` APIs.
> * Creating a `FileDescriptor` and using `InputStream`.
> 
> Access through file path will STILL give permission error (on any SDK level higher than 29, Android 10).
> 
> Similar instances may be found at: https://stackoverflow.com/questions/47125837/java-io-filenotfoundexception-from-intent-data-tostring.

I am getting your point , that is why I tried explicitly asking permissions (READ_MEDIA_IMAGES ..etc) for devices running 13.
But what am I missing here is why there is no issue with my app. Its about file-transfer so it has to get all the paths for files that user selects. Till now I haven't faced any issue. It basically streams files using http. Actually I raised the issue about caching because I couldn't retrieve paths for large files instantly. So I had to tweak file-picker. 
This is the app and you can check it out.

[Photon -file transfer](https://github.com/abhi16180/photon/releases)

**Comment by MSKoWa** [comment]
Hello there,

I am not quite sure if this is the correct issue because I did not get all of the previous comments but the title seems to fit. I am facing the problem Android 13 that I cannot select PDF files from the users' directories. The use case is just some PDF file upload to some server; the PDFs are generated by a third party.

I introduced the permissions for images, videos and audio files in Android 13 and I also have the "READ_EXTERNAL_STORAGE" entry for older SDKs.

However, whenever a user tries to open the file browser with the "file_picker" library on Android 13 devices, there is just no permission to do that (e. g. picking the pdf file) and the user cannot select the PDF due to no file browser. Any advice on this? Thanks in advance!

Best,
Michael
